client/src/rpc_client.rs

```diff
@@ -25,12 +26,187 @@ use solana_sdk::{
     signature::{KeypairUtil, Signature},
     transaction::{self, Transaction, TransactionError},
 };
+use std::sync::atomic::{AtomicBool, Ordering};
+use std::sync::mpsc::{channel, Receiver};
+use std::sync::{Arc, RwLock};
+use std::thread::JoinHandle;
 use std::{
     error, io,
     net::SocketAddr,
     thread::sleep,
     time::{Duration, Instant},
 };
+use thiserror::Error;
+use tungstenite::{client::AutoStream, connect, Message, WebSocket};
+use url::{ParseError, Url};
+
+#[derive(Debug, Error)]
+pub enum PubsubClientError {
+    #[error("url parse error")]
+    UrlParseError(#[from] ParseError),
+
+    #[error("unable to connect to server")]
+    ConnectionError(#[from] tungstenite::Error),
+
+    #[error("json parse error")]
+    JsonParseError(#[from] serde_json::error::Error),
+
+    #[error("unexpected message format")]
+    UnexpectedMessageError,
+}
+
+#[derive(Serialize, Deserialize, PartialEq, Clone, Debug)]
+pub struct SlotInfoMessage {
+    pub parent: u64,
+    pub root: u64,
+    pub slot: u64,
+}
+
+pub struct PubsubClientSlotSubscription {
+    socket: Arc<RwLock<WebSocket<AutoStream>>>,
+    subscription_id: u64,
+    t_cleanup: Option<JoinHandle<()>>,
+    exit: Arc<AtomicBool>,
+}
+
+impl Drop for PubsubClientSlotSubscription {
+    fn drop(&mut self) {
+        self.send_unsubscribe()
+            .unwrap_or_else(|_| warn!("unable to unsubscribe from websocket"));
+        self.socket
+            .write()
+            .unwrap()
+            .close(None)
+            .unwrap_or_else(|_| warn!("unable to close websocket"));
+    }
+}
+
+impl PubsubClientSlotSubscription {
+    fn send_subscribe(
+        writable_socket: &Arc<RwLock<WebSocket<AutoStream>>>,
+    ) -> std::result::Result<u64, PubsubClientError> {
```

Comment:
You have several usages of `std::result::Result` in this file, which I think are just adding unnecessary clutter. Just `Result` should be fine.